diff --git a/dwl.c b/dwl.c
--- a/dwl.c
+++ b/dwl.c
@@ -1796,6 +1812,11 @@ mapnotify(struct wl_listener *listener, void *data)
 
 unset_fullscreen:
 	m = c->mon ? c->mon : xytomon(c->geom.x, c->geom.y);
+	if (c->isfloating && !c->isfullscreen && !client_is_unmanaged(c)
+			&& !client_get_parent(c))
+		resize(c, (struct wlr_box){.x = (m->w.width - c->geom.width) / 2 + m->w.x,
+				.y = (m->w.height - c->geom.height) / 2 + m->w.y,
+				.width = c->geom.width, .height = c->geom.height}, 0);
 	wl_list_for_each(w, &clients, link) {
 		if (w != c && w != p && w->isfullscreen && m == w->mon && (w->tags & c->tags))
 			setfullscreen(w, 0);
@@ -2414,6 +2487,8 @@ setmon(Client *c, Monitor *m, uint32_t newtags)
 		/* Make sure window actually overlaps with the monitor */
 		resize(c, c->geom, 0);
 		c->tags = newtags ? newtags : m->tagset[m->seltags]; /* assign tags of target monitor */
+		c->prev.x = (m->w.width - c->prev.width) / 2 + m->m.x;
+		c->prev.y = (m->w.height - c->prev.height) / 2 + m->m.y;
 		setfullscreen(c, c->isfullscreen); /* This will call arrange(c->mon) */
 		setfloating(c, c->isfloating);
 	}
