diff --git a/dwl.c b/dwl.c
--- a/dwl.c
+++ b/dwl.c
@@ -1779,7 +1794,8 @@ mapnotify(struct wl_listener *listener, void *data)
 	c->geom.height += 2 * c->bw;
 
 	/* Insert this client into client lists. */
-	wl_list_insert(&clients, &c->link);
+	/* tile at the bottom: insert before the list's last element */
+	wl_list_insert(clients.prev, &c->link);
 	wl_list_insert(&fstack, &c->flink);
 
 	/* Set initial monitor, tags, floating status, and focus:
